### ISSUE TITLE
Sending sparse tensors over RPC fails

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#30807]
## To Reproduce

```
    @dist_init
    def test_sparse_tensors(self):
        t1 = torch.rand((3, 3), requires_grad=True).to_sparse()
        t2 = torch.rand((3, 3), requires_grad=True).to_sparse()
        rpc.rpc_sync("worker{}".format(self._next_rank()), torch.add, args=(t1, t2))

```

This fails with the following stack trace (need to catch exceptions in ProcessGroupAgent::enqueueSend to see this):

```
sparse tensors do not have storage (storage at caffe2/aten/src/ATen/SparseTensorImpl.cpp:87)
> frame #0: <unknown function> 
> frame #1: <unknown function> 
> frame #2: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const
> frame #3: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> frame #4: at::SparseTensorImpl::storage() const
> frame #5: <unknown function>
> frame #6: torch::jit::Pickler::pushStorageOfTensor(at::Tensor const&) 
> frame #7: torch::jit::Pickler::pushLiteralTensor(c10::IValue const&)
> frame #8: torch::jit::Pickler::pushTensor(c10::IValue const&)
> frame #9: torch::jit::Pickler::pushIValueImpl(c10::IValue const&) 
> frame #10: torch::jit::Pickler::pushIValue(c10::IValue const&)
> frame #11: <unknown function>
> frame #12: <unknown function>
> frame #13: std::function<void (c10::IValue const&)>::operator()(c10::IValue const&) const
> frame #14: torch::jit::Pickler::pushSpecializedList(c10::IValue const&, torch::jit::PicklerClass, std::function<void (c10::IValue const&)> const&)
> frame #15: torch::jit::Pickler::pushIValueImpl(c10::IValue const&)
> frame #16: torch::jit::Pickler::pushIValue(c10::IValue const&)
> frame #17: torch::distributed::rpc::wireSerialize[abi:cxx11](std::vector<char, std::allocator<char> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&)
> frame #18: <unknown function> 
> frame #19: <unknown function>
> frame #20: <unknown function>
> frame #21: <unknown function> 
> frame #22: <unknown function> 
> frame #23: <unknown function>
> frame #24: std::function<void ()>::operator()() const 
> frame #25: c10::ThreadPool::main_loop(unsigned long) 
> frame #26: <unknown function> 
> frame #27: <unknown function> 
> frame #28: <unknown function> 
> frame #29: <unknown function> 
> frame #30: <unknown function> 
> frame #31: <unknown function> 
> frame #32: <unknown function> 
> frame #33: <unknown function> 
> frame #34: clone 
```

**Comment by pritamdamania87** [comment]
@suo Are sparse tensors not supported by the jit pickler currently? We're using the jit pickler in the RPC framework and it looks like we run into this issue.

**Comment by zhaojuanmao** [comment]
@pritamdamania87 this failed at process group serialization layer that is using the function wireSerialize() newly added and it is added to replace of torch::save(), not the message serialization layer that uses jit::pickler directly and looks like message serialization layer is fine?

do you want to try whether it still fail for sparse tensors using the previous commit that uses torch::save()?

**Comment by pritamdamania87** [comment]
cc @jjlilley

**Comment by pritamdamania87** [comment]
> @pritamdamania87 this failed at process group serialization layer that is using the function wireSerialize() newly added and it is added to replace of torch::save(), not the message serialization layer that uses jit::pickler directly and looks like message serialization layer is fine?

> do you want to try whether it still fail for sparse tensors using the previous commit that uses torch::save()?


I think we should fix this in the current version since the previous version with `torch::save` was much slower.

**Comment by zhaojuanmao** [comment]
agree, the test is just to confirm the root cause is because of the new serializer

**Comment by jjlilley** [comment]
The previous torch::save() approach doesn't support serializing sparse tensors either, here's a quick test case that you can run:

{
   at::Tensor tensor =                                                                                                                                   
      at::empty({2, 3}, at::dtype<float>().layout(at::kSparse));                                                                                         
   EXPECT_TRUE(tensor.is_sparse());                                                                                                                      
                                                                                                                                                         
   std::string str;                                                                                                                                      
   torch::save({tensor}, [&](const void* buf, size_t sz) -> size_t {                                                                                     
                           str.append(static_cast<const char*>(buf), sz);                                                                                
                           return sz;                                                                                                                    
                         });                                                                                                                             
   std::vector<torch::Tensor> tout;                                                                                                                      
   torch::load(tout, str.data(), str.size());                                                                                                            
   EXPECT_EQ(tout.size(), 1);                                                                                                                            
   EXPECT_TRUE(torch::equal(tout[0], tensor));                                                                                                           
   EXPECT_TRUE(tout[0].is_sparse());                                                                                                                     
}

Currently, the output is:
C++ exception with description "sparse tensors do not have storage (storage at caffe2/aten/src/ATen/SparseTensorImpl.cpp:87)

So, reverting to the old torch::save() serialization approach won't help in this regard - sparse Tensors apparently simply aren't supported by the pickler currently.

(fwiw, the new format is simply a stripped down version of torch::save, but they both use the same core Pickler logic for the heavy lifting; as an exercise, compare vs. ScriptModuleSerializer::writeArchive, the new code is mostly a stripped down version)

Should we then check with the pickler folks, about plans for supporting sparse tensors?

**Comment by pritamdamania87** [comment]
@suo Could you assign this to the right person? For more context, we didn't run into this issue in some of our production workloads, but I ran into this while writing a unit test and in general it would be good to support this. I can imagine users trying to send sparse tensors over RPC. 

cc @driazati

**Comment by suo** [comment]
I believe @bwasti is the right POC for sparse tensor-y things.

**Comment by gchanan** [comment]
I don't think non-JIT serialization supports this either.

**Comment by gcramer23** [comment]
I added support for sending CPU tensors over RPC.

This pr is for CUDA tensors https://github.com/pytorch/pytorch/pull/62960. It was backed out because it causes flaky tests https://github.com/pytorch/pytorch/issues/64518.
